landoapi/uplift.py

```diff
@@ -22,6 +23,48 @@
 logger = logging.getLogger(__name__)
 
 
+ARC_DIFF_REV_RE = re.compile(
+    r"^\s*Differential Revision:\s*(?P<phab_url>https?://.+)/D(?P<rev>\d+)\s*$",
+    flags=re.MULTILINE,
+)
+ORIGINAL_DIFF_REV_RE = re.compile(
+    r"^\s*Original Revision:\s*(?P<phab_url>https?://.+)/D(?P<rev>\d+)\s*$",
+    flags=re.MULTILINE,
+)
+
+
+def move_drev_to_original(body: str) -> str:
+    """Handle moving the `Differential Revision` line.
+
+    Moves the `Differential Revision` line to `Original Revision`, if a link
+    to the original revision does not already exist. If the `Original Revision`
+    line does exist, scrub the `Differential Revision` line.
+
+    Args:
+        body: `str` text of the commit message.
+        rev_id: `int` parsed integer representing the drev number for the revision.
+
+    Returns:
+        tuple of:
+            New commit message body text as `str`,
+            Revision id as `int`, or `None` if a new revision should be created.
+    """
+    differential_revision = ARC_DIFF_REV_RE.search(body)
+    original_revision = ORIGINAL_DIFF_REV_RE.search(body)
+
+    # If both match, we already have an `Original Revision` line.
+    if differential_revision and original_revision:
+        return body
+
+    def repl(match):
+        phab_url = match.group("phab_url")
+        rev = match.group("rev")
+        return f"\nOriginal Revision: {phab_url}/D{rev}"
```

Comment:
Could be a little cleaner like this:

```suggestion
        return "\nOriginal Revision: {phab_url}/D{rev}".format(**match.groupdict())
```

Or perhaps `repl` is not necessary as a method.

Reply:
We already have this code in `moz-phab`: https://github.com/mozilla-conduit/review/blob/main/mozphab/helpers.py#L219

I copied this over to keep the logic for moving `Differential Revision` consistent between lando and moz-phab, but removed the extra `rev_id` parsing that we have in moz-phab. I'm probably going to carry that into `moz-phab` and have the `rev_id` parsing take place elsewhere.

`repl` is necessary for `re.sub`, and "substituting `Differential Revision` for `Original Revision`" is the plain-english description of what we're trying to achieve.

Moving it into a one-liner is a -1 from me, it doesn't make anything clearer and makes the regex groups used more implicit instead of explicit. It removes a few lines, but disk space is cheap. ;)

Reply:
Should have prefixed this with "nit", and I won't block the PR on it.

I think naming these groups without using `groupdict` doesn't utilize these named groups to their full potential given what you're using it for. Since the keys, group names, and f-string placeholders (i.e. variable names) all match, this seems repetitive and less readable. A possible middle ground:

```suggestion
        result = match.groupdict()
        return f"\nOriginal Revision: {result['phab_url']}/D{result['rev']}"
```

But again, this is probably splitting hairs at this point.